src/adapter/src/client.rs

```diff
@@ -811,30 +811,37 @@ impl SessionClient {
             cmd
         });
 
+        // SPOOKY ZONE! See the doc comment on `SessionTerminateGuard` for why this exists.
+        //
+        // tl;dr this guard prevents a race where a `Session` is returned on `rx` but never placed
+        // back in `self` because the Future returned by this function is concurrently dropped.
+        let mut guard = SessionTerminateGuard { client: self, rx };
+
         let mut cancel_future = pin::pin!(cancel_future);
         let mut cancelled = false;
         loop {
             tokio::select! {
-                res = &mut rx => {
+                res = &mut guard => {
                     let res = res.expect("sender dropped");
                     let status = if res.result.is_ok() {
                         "success"
                     } else {
                         "error"
                     };
                     if let Some(typ) = typ {
-                        self.inner()
+                        guard.client
+                            .inner()
                             .metrics
                             .commands
                             .with_label_values(&[typ, status, name_hint.as_str()])
                             .inc();
                     }
-                    self.session = Some(res.session);
+                    guard.client.session = Some(res.session);
                     return res.result
                 },
                 _err = &mut cancel_future, if !cancelled => {
```

Comment:
While we are here, we should add docs saying that the `cancel_safe` future can be cancelled partially completed (actually, is this another bug?)

Reply:
Sure, probably worth documenting on `execute` that `cancel_future` must be cancel safe. It's not a bug AFAICT because the one future that we pass in here is in fact cancel safe.

Reply:
Added!